src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt

```diff
@@ -1,52 +1,126 @@
+        queryBy(searchCriteria) { nameMatchWithAliases(it) } onMatch { return it.matches }
+      } else {
+        queryBy(searchCriteria) { nameMatch(it) } onMatch { return it.matches }
+      }
+    }
+    return emptyList()
   }
 
-  fun findByKeywords(keywords: String, pageable: Pageable): Page<Prisoner> {
-    return getPrisonerRepository().findByKeywords(keywords, pageable)
+  private fun validateSearchForm(searchCriteria: SearchCriteria) {
+    if (!searchCriteria.isValid) {
+      log.warn("Invalid search  - no criteria provided")
+      throw BadRequestException("Invalid search  - please provide at least 1 search parameter")
+    }
   }
 
-  fun findByKeywordsFilterByPrison(keywords: String, prisonId : String, pageable: Pageable): Page<Prisoner> {
-    return getPrisonerRepository().findByKeywordsFilterByPrison(keywords, prisonId, pageable)
+  private fun queryBy(searchCriteria: SearchCriteria, queryBuilder: (searchCriteria: SearchCriteria) -> BoolQueryBuilder?): Result {
+    val query = queryBuilder(searchCriteria)
+    return query?.let {
+      val searchSourceBuilder = SearchSourceBuilder().apply {
+        query(query.withDefaults(searchCriteria))
+      }
+      val searchRequest = SearchRequest(arrayOf(getIndex()), searchSourceBuilder)
+      val prisonerMatches = getSearchResult(searchClient.search(searchRequest))
+      return if (prisonerMatches.isEmpty()) Result.NoMatch else Result.Match(prisonerMatches)
+    } ?: Result.NoMatch
   }
 
-  fun findByDob(dob: LocalDate, pageable: Pageable): Page<Prisoner> {
-    return getPrisonerRepository().findByDateOfBirth(dob, pageable)
+  private fun idMatch(searchCriteria: SearchCriteria): BoolQueryBuilder? {
+    with(searchCriteria) {
+      return QueryBuilders.boolQuery()
+        .mustMultiMatchKeyword(prisonerIdentifier?.canonicalPNCNumber(), "prisonerNumber", "bookingId", "pncNumber", "croNumber", "bookNumber")
+    }
   }
 
-  fun findByBookingId(bookingId: Long): Prisoner? {
-    return getPrisonerRepository().findByBookingId(bookingId)
+  private fun nameMatch(searchCriteria: SearchCriteria): BoolQueryBuilder? {
+    with(searchCriteria) {
+      return QueryBuilders.boolQuery()
+        .must(QueryBuilders.boolQuery()
+          .should(QueryBuilders.boolQuery()
+            .mustWhenPresent("lastName", lastName)
+            .mustWhenPresent("firstName", firstName)
+          )
+        )
+    }
   }
 
-  fun findByPrisonId(prisonId: String, pageable: Pageable): Page<Prisoner> {
-    return getPrisonerRepository().findByPrisonId(prisonId, pageable)
+  private fun nameMatchWithAliases(searchCriteria: SearchCriteria): BoolQueryBuilder? {
+    with(searchCriteria) {
+      return QueryBuilders.boolQuery()
+        .must(QueryBuilders.boolQuery()
+          .should(QueryBuilders.boolQuery()
+            .mustWhenPresent("lastName", lastName)
+            .mustWhenPresent("firstName", firstName)
+          )
+          .should(QueryBuilders.nestedQuery("aliases",
```

Comment:
Interestingly I think I have a bug in probation search where I am not doing a nested search,  I think I need to raise a bug ticket

Reply:
@andymarke yes without nesting the relationship between array entries is lost, ie it will match for a first name in one record with a last name in another